confmap/provider_test.go

```diff
@@ -6,13 +6,87 @@ package confmap
 import (
 	"context"
 	"errors"
+	"fmt"
 	"testing"
 	"time"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
+// This is an example of a provider that calls a provided WatcherFunc to update configuration dynamically every second.
+// The example is useful for implementing Providers of configuration that changes over time.
+type UpdatingProvider struct{}
+
+func (p UpdatingProvider) getCurrentConfig(_ string) any {
+	return "hello"
+}
+
+func (p UpdatingProvider) Retrieve(ctx context.Context, uri string, watcher WatcherFunc) (*Retrieved, error) {
+	ticker := time.NewTicker(1 * time.Second)
+	stop := make(chan bool, 1)
+
+	retrieved, err := NewRetrieved(p.getCurrentConfig(uri), WithRetrievedClose(func(_ context.Context) error {
+		// the provider should call this function when it no longer wants config updates
+		ticker.Stop()
+		stop <- true
+		return nil
+	}))
+	if err != nil {
+		return nil, err
+	}
+	// it's necessary to start a go function that can notify the caller of changes asynchronously
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				// if the context is closed, then we should stop sending updates
+				ticker.Stop()
+				return
+			case <-stop:
+				// closeFunc was called, so stop updating the watcher
+				fmt.Println("unsubscribing watcher")
+				return
+			case <-ticker.C:
+				// the configuration has "changed". Notify the watcher that a new config is available
+				// the watcher is expected to call Provider.Retrieve again to get the update
+				// note that the collector calls closeFunc before calling Retrieve for the second time,
+				// so these go functions don't accumulate indefinitely. (see otelcol/collector.go, Collector.reloadConfiguration)
+				watcher(&ChangeEvent{})
+			}
+		}
+	}()
+
+	return retrieved, nil
+}
+
+func ExampleProvider() {
```

Comment:
```suggestion
// Tests the example scenario in README.md
func TestProviderWithWatcher(t *testing.T) {
```
Could we make this into a test and verify the output specified in the "Output" comment section of this function?

Reply:
@mx-psi  wanted this to be a testable example

Reply:
@evan-bradley we don't have to change the name (we can keep it as an example) to use the "Output comment", right?

Reply:
Maybe I misunderstood the intent of this, I guess what has me confused is that this is in the test file despite not being a test. Should we move it to the `doc.go` file or an internal `examples` directory instead?

Reply:
> I guess what has me confused is that this is in the test file despite not being a test

Go allows you to write down [testable examples](https://go.dev/blog/examples) if you start the function name with `Example`. This is the thing that powers https://pkg.go.dev/net/http#pkg-examples, it shows up nicely in the Godoc

Reply:
Oh wow, I didn't know about that, that's a cool feature. Sorry, just now seeing your earlier comment where you linked to this earlier. I think what's here is perfect then.